internal/route/websocket.go

```diff
@@ -0,0 +1,113 @@
+package route
+
+import (
+	"log"
+	"net/http"
+
+	"github.com/gorilla/websocket"
+)
+
+// Message that will be sent to client
+type Message struct {
+	Event string `json:"event"`
+	Data  struct {
+		ListID string `json:"list_id"`
+		TodoID string `json:"todo_id"`
+		Done   bool   `json:"done"`
+	} `json:"data"`
+}
+
+// RegisterMessage registers client
+type RegisterMessage struct {
+	Message Message
+	Client  *websocket.Conn
+}
+
+type socketManager struct {
+	clients   map[string][]*websocket.Conn
+	register  chan RegisterMessage
+	broadcast chan Message
+}
+
+var manager = socketManager{
+	clients:   make(map[string][]*websocket.Conn),
+	register:  make(chan RegisterMessage), // registration channel
+	broadcast: make(chan Message),         // broadcast channel
+}
+
+var upgrader = websocket.Upgrader{}
+
+// HandleWs handles websocket connections
+func HandleWs(w http.ResponseWriter, r *http.Request) {
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer ws.Close()
+
+	for {
+		var msg Message
+		err := ws.ReadJSON(&msg)
+
+		if err != nil {
+			log.Printf("error: %v\n", err)
+			break
+		}
+
+		if msg.Event == "register" {
+			registerMsg := RegisterMessage{
+				Message: msg,
+				Client:  ws,
+			}
+			manager.register <- registerMsg
+		} else {
+			manager.broadcast <- msg
+		}
+	}
+}
+
+// HandleMessages processes WS messages and sends them to the client
+func HandleMessages() {
+	for {
+		select {
+		case msg := <-manager.broadcast:
+			manager.handleMessage(&msg)
+		case reg := <-manager.register:
+			manager.registerClient(&reg)
+		}
+	}
+}
+
+func (m *socketManager) removeClient(listID string, c *websocket.Conn) {
+	listClients := m.clients[listID]
+	for i, client := range listClients {
+		if c == client {
+			m.clients[listID] = append(m.clients[listID][:i], m.clients[listID][i+1:]...)
+			break
+		}
+	}
+}
```

Comment:
FWIW: I'd try to avoid two things here:

1. Having a for / if / true-code case. It's three levels of indentation (on top of the func) so it's really moving right.

2. Using a loop to do a filter. A functional approach prevents having to loop and break and creates a cleaner flow. Here's how I'd approach this:

```go

func filterConn(conns []*websocket.Conn, c *websocket.Conn) (int, *websocket.Conn) {
    for i, curconn := range conns {
        if curconn == c {
            return i, curconn
        }
    }
    return -1, nil
}

func (m *socketManager) removeClient(listID string, c *websocket.Conn) {
    listClients := m.clients[listID]
    i, _ := filterConn(listClients, c)
    if c != nil {
        m.clients[listID] = append(m.clients[listID][:i], m.clients[listID][i+1:]...)
    }
}
```